apps/server/lib/lexical/server/provider/handlers/prepare_rename.ex

```diff
@@ -0,0 +1,40 @@
+defmodule Lexical.Server.Provider.Handlers.PrepareRename do
+  alias Lexical.Ast
+  alias Lexical.Document
+  alias Lexical.Protocol.Requests.PrepareRename
+  alias Lexical.Protocol.Responses
+  alias Lexical.Protocol.Types.PrepareRenameResult.PrepareRenameResult
+  alias Lexical.RemoteControl.Api
+  alias Lexical.Server.Provider.Env
+
+  def handle(%PrepareRename{} = request, %Env{} = env) do
+    case Document.Store.fetch(request.document.uri, :analysis) do
+      {:ok, _document, %Ast.Analysis{valid?: true} = analysis} ->
+        prepare_rename(env.project, analysis, request.position, request.id)
+
+      _ ->
+        {:reply,
+         Responses.PrepareRename.error(
+           request.id,
+           :request_failed,
+           "document can not be analyzed"
+         )}
+    end
+  end
+
+  defp prepare_rename(project, analysis, position, id) do
+    case Api.prepare_rename(project, analysis, position) do
+      {:ok, cursor_entity, range} ->
+        result =
+          PrepareRenameResult.new(
+            placeholder: cursor_entity,
+            range: range
+          )
+
+        {:reply, Responses.PrepareRename.new(id, result)}
+
+      _ ->
+        {:reply, Responses.PrepareRename.new(id, nil)}
```

Comment:
if this is the error case, should we respond with a message like the docs say to do?

Reply:
If I understand the document correctly, then for modules that are not at the defined location, we should return null, while for other entities, we should return an error message.

Reply:
Yes

Reply:
Updated. Please take a look at the latest commit.

![CleanShot 2024-06-20 at 14 02 26@2x](https://github.com/lexical-lsp/lexical/assets/12830256/3de333b7-68b8-42c5-b48f-ea99a91a2cea)